pkg/gateway/api/grpc_validation_test.go

```diff
@@ -0,0 +1,478 @@
+// Copyright 2022 The Bucketeer Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package api
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes/any"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/bucketeer-io/bucketeer/pkg/log"
+	eventproto "github.com/bucketeer-io/bucketeer/proto/event/client"
+	"github.com/bucketeer-io/bucketeer/proto/feature"
+)
+
+const (
+	oldestTimestampDuration   = 24 * time.Hour
+	furthestTimestampDuration = 24 * time.Hour
+)
+
+func TestNewEventValidator(t *testing.T) {
+	t.Parallel()
+	bEvaluationEvent, err := proto.Marshal(&eventproto.EvaluationEvent{})
+	if err != nil {
+		t.Fatal("could not serialize evaluation event")
+	}
+	bGoalEvent, err := proto.Marshal(&eventproto.GoalEvent{})
+	if err != nil {
+		t.Fatal("could not serialize goal event")
+	}
+	bGoalBatchEvent, err := proto.Marshal(&eventproto.GoalBatchEvent{})
+	if err != nil {
+		t.Fatal("could not serialize goal batch event")
+	}
+	bMetricsEvent, err := proto.Marshal(&eventproto.MetricsEvent{})
+	if err != nil {
+		t.Fatal("could not serialize metrics event")
+	}
+	patterns := map[string]struct {
```

Comment:
> Because there are many tests using both, we decided not to use a map to maintain consistency.

It doesn't make sense to me. It's no reason for using an array, too.

Reply:
Could you explain why a map makes sense to you?

Reply:
> Btw, I prefer the field name as name instead of desc like [pipecd](https://github.com/pipe-cd/pipecd/blob/master/test/integration/filestore/gcs/gcs_test.go#L65-L70).

At least that's not the out of scope in this PR. If you wanna change them, please create another PR and propose it.

Reply:
>Could you explain why a map makes sense to you?

Because they are used in the following files:
* pkg/account/api/account_test.go
* pkg/account/api/admin_account_test.go
* pkg/account/api/api_key_test.go
* ...

Reply:
I'm asking why it makes sense to you. Listing places that are using a map is not a good reason.
Also, as I described at the beginning of this conversation, I said that we have both implementations but decided to use an array.